File: Sources/iStream/Models/PipelineDesign.swift
import Foundation

public struct TopicConfig: Equatable, Sendable {
    public var name: String
    public var partitions: Int
    public var replicationFactor: Int
    public var minInSyncReplicas: Int
    public var retentionMs: Int64
    public var cleanupPolicy: String
    public var compressionType: String
}

public struct FlinkConfig: Equatable, Sendable {
    public var parallelism: Int
    public var stateBackend: String
    public var checkpointIntervalMs: Int
    public var checkpointMode: String
    public var watermarkBoundMs: Int
    public var allowedLatenessMs: Int
}

public struct PipelineDesign: Identifiable, Equatable, Sendable {
    public let id: UUID
    public var input: PipelineInput
    public var topic: TopicConfig
    public var flink: FlinkConfig
    public var skewMitigation: [String]
    public var notes: [String]
    public var mermaid: String
    public var markdown: String

    public init(
        id: UUID = UUID(),
        input: PipelineInput,
        topic: TopicConfig,
        flink: FlinkConfig,
        skewMitigation: [String],
        notes: [String],
        mermaid: String,
        markdown: String
    ) {
        self.id = id
        self.input = input
        self.topic = topic
        self.flink = flink
        self.skewMitigation = skewMitigation
        self.notes = notes
        self.mermaid = mermaid
        self.markdown = markdown
    }
}

File: lib/models/pipeline_design.dart
import 'pipeline_input.dart';

class TopicConfig {
  const TopicConfig({
    required this.name,
    required this.partitions,
    required this.replicationFactor,
    required this.minInSyncReplicas,
    required this.retentionMs,
    required this.cleanupPolicy,
    required this.compressionType,
  });

  final String name;
  final int partitions;
  final int replicationFactor;
  final int minInSyncReplicas;
  final int retentionMs;
  final String cleanupPolicy;
  final String compressionType;
}

class FlinkConfig {
  const FlinkConfig({
    required this.parallelism,
    required this.stateBackend,
    required this.checkpointIntervalMs,
    required this.checkpointMode,
    required this.watermarkBoundMs,
    required this.allowedLatenessMs,
  });

  final int parallelism;
  final String stateBackend;
  final int checkpointIntervalMs;
  final String checkpointMode;
  final int watermarkBoundMs;
  final int allowedLatenessMs;
}

class PipelineDesign {
  const PipelineDesign({
    required this.input,
    required this.topic,
    required this.flink,
    required this.skewMitigation,
    required this.notes,
    required this.mermaid,
    required this.markdown,
  });

  final PipelineInput input;
  final TopicConfig topic;
  final FlinkConfig flink;
  final List<String> skewMitigation;
  final List<String> notes;
  final String mermaid;
  final String markdown;
}
